test/test_meta.py

```diff
@@ -66,6 +66,34 @@ def assertSameVersionCounter(self, m1, m2):
         self.assertNotEqual(m1._version, vc)
         self.assertEqual(m2._version, m1._version)
 
+    def assertMetadataMatches(self, m1, m2):
+        self.assertEqual(m1.dtype, m2.dtype)
+        self.assertEqual(m1.shape, m2.shape)
+        self.assertEqual(m1.requires_grad, m2.requires_grad)
+        self.assertEqual(m1.is_leaf, m2.is_leaf)
+        self.assertEqual(m1.grad_fn is None, m2.grad_fn is None)
+        self.assertEqual(m1.is_sparse, m2.is_sparse)
+        self.assertEqual(m1.is_inference(), m2.is_inference())
+        self.assertEqual(m1.is_conj(), m2.is_conj())
+        self.assertEqual(m1.is_neg(), m2.is_neg())
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore", "The .grad attribute of a Tensor")
+            grad_not_none = m1.grad is not None
+        if grad_not_none:
+            self.assertMetadataMatches(m1.grad, m2.grad)
+        if m1.is_sparse:
+            self.assertEqual(m1.dense_dim(), m2.dense_dim())
+            self.assertEqual(m1.sparse_dim(), m2.sparse_dim())
+            self.assertEqual(m1.is_coalesced(), m2.is_coalesced())
+        else:
+            self.assertEqual(m1.stride(), m2.stride())
+            self.assertEqual(m1.storage_offset(), m2.storage_offset())
+            self.assertEqual(m1._is_view(), m2._is_view())
+            if m1._is_view():
+                self.assertMetadataMatches(m1._base, m2._base)
+        # TODO: test if is resizable (no direct query for this atm)
+        # TODO: audit AutogradMeta to see if it matches
```

Comment:
Forward AD agreed.

Hooks, I'm not so sure. Can a hook trigger if it is attached to the INPUT tensor of a grad compute? I guess if it's a pre-hook? But it will trigger anyway when the gradient flows to the original, real tensor. So you might double trigger the hook in that situation.